### ISSUE TITLE
Apple Shortcuts support

### ISSUE THREAD

**Issue by matthinc** [nextcloud/ios#662]
A feature that I would love to see:

Apple introduced a „Shortcuts“ App for ios12 which allows to build fairly complex automations. Nextcloud integration would be awesome. 
You could build for example:
- Share file to automation
- Upload to nextcloud
- Share file via link
- Send link to friend via messages...

The possibilities would be endless.

**Comment by marinofaggiana** [comment]
Hi @matthinc, sure in the future this will implemented.

**Comment by melikyuksel** [comment]
Any word on this? Would help people to make their own features, as the app development seems rather behind.

**Comment by normen** [comment]
It's especially sad that when using the "save file" option in iOS Shortcuts you cannot save directly to NextCloud, only to iCloud or DropBox. However when letting the user choose a save location the user can navigate to the NextCloud folder... 🤷

**Comment by melyux** [comment]
Save to Files doesn’t work anyway. Has never worked. Stuck “In upload” every time. I’ve given up on Nextcloud because of it.

**Comment by normen** [comment]
Nah, works fine for me.

**Comment by normen** [comment]
Btw, heres a shortcut that uses WebDav to be able upload directly to NextCloud at least: https://www.reddit.com/r/NextCloud/comments/e5e1bs/ios_shortcut_to_upload_file_and_copy_link_to/

**Comment by GitGramm** [comment]
owncloud provides a feature like this: https://owncloud.com/features/apple-shortcuts/ 
would love to see something similar implemented in the nextcloud ios app

**Comment by sneibel** [comment]
Yes please, the integration of ios shortcuts is the only feature I wait for to drop dropbox as my cloud storage and to use nextcloud on my own server.

**Comment by xplosionmind** [comment]
Something I would love is to be able to directly edit files, for example being able to add a CSV row as a result for a shortcut.

**Comment by SommerEngineering** [comment]
Is there an update on this issue? iOS shortcuts are a key feature for power users. In particular, I love the idea of triggering photo uploads with a shortcut. The idea of attaching a new line to a text file (e.g. CSV or Markdown) is also brilliant for logging events and analyzing them on another machine. A third important command would be that we can read and iterate the contents of directories in a shortcut.

Just if I may make an interface suggestion from a shortcut user perspective:

- Action: "Photo Sync"
   - Input: just a trigger; no files etc.
   - Output: message as string ("successful" or any error message)
   - Behavior: when triggered, tries to sync the iOS photos to the nextcloud server. Blocks until finished.
   - Use case: triggering photo sync on a well-known fast WiFi

- Action: "Append Lines"
   - Input: one string or a list of strings (each list entry is one line) and a filename (path)
   - Output: message as string ("successful" or any error message)
   - Behavior: tries to append the string (one line) or the list of strings (multiple lines) to the given file. When the file does not exists, create the file. Each line ends with `\n`. May, let the user choose the desired line-ending (Linux vs. Windows). The action should block until finished.
   - Use case: maintain a journal i.e. markdown note or CSV file with observations. For example, iOS HomeKit (Apple's smart home integration) could trigger, when my security camera detects motion in front of my door; in such a event, we could write the date and time to such a text file and upload a photo as well. Second example: by using an iOS shortcut, you create a geo-fence to detect your workplace or the locations of your clients. The "append lines" would enable us to log when we arrive and leave such a location. Based on that data, we could write an invoice for our clients etc. 

- Action: "List Directory Content"
   - Input: a path as string
   - Output: a list of strings representing the files (absolute paths); may let the user choose wherever the operation should list recursive all the sub-directories as well; let the user choose wherever hidden files should be listed as well; might, the user could also add a pattern like `*.md` to get only such files.
   - Behavior: returns a list of absolute filenames of the directory.
   - Use case: wait for and consume files, e.g. another computer creates a new file with links or notes etc. and an iOS shortcut can read and present that new file to the user.


These actions might be the most important. In the midterm, we could have more for deeper automation, e.g.:

- Action: "Share a file"
   - Input: path to an existing file in nextcloud; let the user specify a password as well;  let the user choose, wherever the strangers can edit, reshare, delete etc.
   - Output: a string with the link of the shared file
   - Behavior: Blocks until the share is established
   - Use case: you are at a conference and takes photos of all the slides; another participant asks you for sharing one particular slide. With a simple iOS shortcut I could do so. Sending a mail might is worse, because the stranger then knows my mail address or the photos are just too big for a mail etc.

- Action: "Share a directory"
   - Input: path to an existing directory in nextcloud; let the user specify a password as well; let the user choose, wherever the strangers can edit, reshare, delete etc.
   - Output: a string with the link of the shared folder
   - Behavior: Blocks until the share is established
   - Use case: you are at a conference and takes photos of all the slides; another participant asks you for sharing these slides. With a simple iOS shortcut I could do so. Sending a mail might is worse, because the stranger then knows my mail address or the photos are just too big for a mail etc.

- Action: "Stop Sharing"
   - Input: path to a shared file or directory
   - Output: message as string ("successful" or any error message)
   - Behavior: stops sharing of the mentioned resource
   - Use case: you are at a conference (see other sharing use cases); you shared all slides and the stranger loads them; now, you want to stop sharing the folder.

- Action: "List all shares"
   - Input: an optional filter like `*.md`
   - Output: a list of strings with absolute paths to shared resources (file and directories)
   - Behavior: lists all active shared resources
   - Use case: to have an iOS shortcut to get a brief overview of all shared stuff.

- Action: "Create Talk"
   - Input: name as string; let the user configure the password
   - Output: the link for strangers as string
   - Behavior: when the talk addon is installed on the server-side, creates a new talk
   - Use case: create a talk on-the-go for you and a stranger

What do you think? :slightly_smiling_face: This would address #1012 as well.

**Comment by xplosionmind** [comment]
> Is there an update on this issue?

+1 😍

**Comment by jancborchardt** [comment]
Would this possibly also be something to link up with regarding the workflow automation topic? @tobiasKaminsky @AndyScherzinger

**Comment by AndyScherzinger** [comment]
Yes, I agree

**Comment by dxcore35** [comment]
Please!!!! This is super important for Home server in 2024 to have!

**Comment by i2h3** [comment]
@SommerEngineering The Nextcloud app integrates with the iOS Files app. Have you considered using standard shortcut actions shipped with iOS? For some of your described examples there are exactly such actions like listing the files within a directory or appending text to a text file. In some parts we would otherwise just reinvent the wheel.

For some suggestions [App Intents](https://developer.apple.com/documentation/appintents) are the right way to go for implementations by now because they not only expose app functionality to shortcuts but also Siri and Spotlight.

**Comment by segdy** [comment]
@i2h3 No they don’t work that’s why we’re all hoping so much for proper/separate implementation. (Otherwise I would happily use them already)

**Comment by i2h3** [comment]
@tobiasKaminsky Did you close this because #3261 is considered the replacement issue? 😮
